src/librustc_builtin_macros/asm.rs

```diff
@@ -0,0 +1,527 @@
+        } else {
+            if !args.named_args.is_empty() || !args.reg_args.is_empty() {
+                let mut err = ecx.struct_span_err(
+                    span,
+                    "positional arguments cannot follow named arguments \
+                     or explicit register arguments",
+                );
+                err.span_label(span, "positional argument");
+                for pos in args.named_args.values() {
+                    err.span_label(args.operands[*pos].1, "named argument");
+                }
+                for pos in &args.reg_args {
+                    err.span_label(args.operands[*pos].1, "explicit register argument");
+                }
+                err.emit();
+            }
+        }
+    }
+
+    if args.options.contains(InlineAsmOptions::NOMEM)
+        && args.options.contains(InlineAsmOptions::READONLY)
+    {
+        let span = args.options_span.unwrap();
+        ecx.struct_span_err(span, "the `nomem` and `readonly` options are mutually exclusive")
+            .emit();
+    }
+    if args.options.contains(InlineAsmOptions::PURE)
+        && args.options.contains(InlineAsmOptions::NORETURN)
+    {
+        let span = args.options_span.unwrap();
+        ecx.struct_span_err(span, "the `pure` and `noreturn` options are mutually exclusive")
+            .emit();
+    }
+    if args.options.contains(InlineAsmOptions::PURE)
+        && !args.options.intersects(InlineAsmOptions::NOMEM | InlineAsmOptions::READONLY)
+    {
+        let span = args.options_span.unwrap();
+        ecx.struct_span_err(
+            span,
+            "the `pure` option must be combined with either `nomem` or `readonly`",
+        )
+        .emit();
+    }
+
+    let mut have_real_output = false;
+    let mut outputs_sp = vec![];
+    for (op, op_sp) in &args.operands {
+        match op {
+            ast::InlineAsmOperand::Out { expr, .. }
+            | ast::InlineAsmOperand::SplitInOut { out_expr: expr, .. } => {
+                outputs_sp.push(*op_sp);
+                have_real_output |= expr.is_some();
+            }
+            ast::InlineAsmOperand::InOut { .. } => {
+                outputs_sp.push(*op_sp);
+                have_real_output = true;
+            }
+            _ => {}
+        }
+    }
+    if args.options.contains(InlineAsmOptions::PURE) && !have_real_output {
+        ecx.struct_span_err(
+            args.options_span.unwrap(),
+            "asm with `pure` option must have at least one output",
+        )
+        .emit();
+    }
+    if args.options.contains(InlineAsmOptions::NORETURN) && !outputs_sp.is_empty() {
+        let err = ecx
+            .struct_span_err(outputs_sp, "asm outputs are not allowed with the `noreturn` option");
+
+        // Bail out now since this is likely to confuse MIR
+        return Err(err);
+    }
+
+    Ok(args)
+}
+
+fn expand_preparsed_asm(ecx: &mut ExtCtxt<'_>, sp: Span, args: AsmArgs) -> P<ast::Expr> {
```

Comment:
**Note to self**: review this later once the question of `rustc_ast` depending on `rustc_target`.